nano/node/vote_processor.cpp

```diff
@@ -88,56 +89,48 @@ void nano::vote_processor::process_loop ()
 	}
 }
 
-void nano::vote_processor::vote (std::shared_ptr<nano::vote> vote_a, std::shared_ptr<nano::transport::channel> channel_a)
+bool nano::vote_processor::vote (std::shared_ptr<nano::vote> vote_a, std::shared_ptr<nano::transport::channel> channel_a)
 {
+	bool process (false);
 	nano::unique_lock<std::mutex> lock (mutex);
 	if (!stopped)
 	{
-		bool process (false);
-		/* Random early delection levels
-		 Always process votes for test network (process = true)
-		 Stop processing with max 144 * 1024 votes */
-		if (!network_params.network.is_test_network ())
+		// Level 0 (< 0.1%)
+		if (votes.size () < 6.0 / 9.0 * max_votes)
 		{
-			// Level 0 (< 0.1%)
-			if (votes.size () < 96 * 1024)
-			{
-				process = true;
-			}
-			// Level 1 (0.1-1%)
-			else if (votes.size () < 112 * 1024)
-			{
-				process = (representatives_1.find (vote_a->account) != representatives_1.end ());
-			}
-			// Level 2 (1-5%)
-			else if (votes.size () < 128 * 1024)
-			{
-				process = (representatives_2.find (vote_a->account) != representatives_2.end ());
-			}
-			// Level 3 (> 5%)
-			else if (votes.size () < 144 * 1024)
-			{
-				process = (representatives_3.find (vote_a->account) != representatives_3.end ());
-			}
+			process = true;
 		}
-		else
+		// Level 1 (0.1-1%)
+		else if (votes.size () < 7.0 / 9.0 * max_votes)
```

Comment:
Not sure if these values can be `static const` to prevent calculation for each vote

Reply:
Static const is pretty bad, could save the values in memory in the class but is memory access better than this? @wezrule what do you think?

Reply:
The easiest gains in this function would be made by passing parameters `vote_a` & `channel_a` by `const ref &` tbh.

Multiplying is pretty cheap, I would personally just do that. I wouldn't use local `static` variables due to the hidden thread synchronization costs which would outweigh it.

However if I was looking to micro-optimize this (as it's a hot path), and profiling showed this to be a bottleneck (I doubt it) then I might consider caching the values as class members, but this would require a lot of thought about cache lines, alignment and data layout; Because by itself it could make the performance worse. Imagine the `vote_processor` is aligned on a 64 byte boundary (typical cache line size) and all data fits neatly into a few cache lines. These extra variables would spill into another one which means if it isn't in the CPU cache already then when this function is called it would require an extra read from main memory (100-200 cycles). During high TPS it will likely be cached already though, as the data would have been recently accessed, but these are the types of things to think about. It would make sense to put all commonly used data members together and organised to reduce padding and make sure they are within the same cache line (align the vote processor on a 64 byte boundary) while you were at it, but any changes to this class in the future would then potentially screw this up (maintenance nightmare).

There's also out of order execution which could make this redundant too. We don't really do this micro-optimisation anywhere else currently, sounds a lot more effort than it's worth at this stage than just multiplying it imo.

Reply:
Thanks for the in-depth details, I agree, having it clear in the code is a larger benefit than the micro optimizations that can be done.